### DISCUSSION TITLE
Performance of v4 vs v3: twice as slow?

### DISCUSSION THREAD

**Discussion by jameshulse** [spatie/laravel-data#713] [General]
Hi team,

I want to preface this discussion by saying that we find this package really useful. We use it extensively now for ViewModels, database model casting, validation, typescript generation. It's a huge help.

But in attempting to upgrade to v4 we ran into a few issues. One key issue that makes me reluctant to upgrade is v4 is that v4 seems to be about twice as slow as v3 in our tests.

We have some endpoints with quite complex (and large) data models. I benchmarked one such endpoint on v3 and v4 and these were the results:

**laravel-data v3 (with structure caching)**
```
Requests per second:    5.20 [#/sec] (mean)
Time per request:       961.319 [ms] (mean)
```

**laravel-data v4 (with structure caching)**
```
Requests per second:    2.53 [#/sec] (mean)
Time per request:       1979.557 [ms] (mean)
```

The results showed that the mean time per request went up significantly post-upgrade. At this kind of performance we will likely have to _not_ use laravel-data for that end-point and opt for raw classes. This is a worry as we may end up with multiple ways of achieving the same thing depending on performance needs. Note: Structure caching didn't seem to make much of a difference to performance, if any.

So I'm wondering: is there any on-going work to improve the performance of the v4 package? Is there anything we can do to help ie. provide xdebug profiles for our endpoint?

**Comment by Tofandel** [comment]
I digged into this and it's coming from
https://github.com/spatie/laravel-data/blob/a5df67e06f1f9012261a9c14d007e81a8fd84046/src/Normalizers/ModelNormalizer.php#L39-L43

Because it loads all of the attributes declared in the model, even if the data class doesn't need them

It was added in v4 https://github.com/spatie/laravel-data/commit/e1e491d20f09b0e07a7199487926c76741d7d39f

So if you have any hungry attribute like this that you don't use in the data class

```php
    public function getSpeakersAttribute(): Arrayable
    {
        $this->loadMissing(['events.video.speakers', 'events.speakers']);

        return $this->events->flatMap(fn (SeminarEvent $event) => $event->speakers)->unique('id')->values();
    }
```

It can take up to 20 times more time to load your models into data class

I made a PR #730 which fixes this issue

**Comment by rubenvanassche** [comment]
Good catch!

**Comment by lentex** [comment]
Thank you for bringing that up (and tracking down the issue). Our company also upgraded to v4 in a project with huge traffic and our PHP docker containers kept crashing. We never found the issue and reverted the changes. At that time I didn't create an issue here because I just couldn't track down the actual issue.

We will give the latest version another shot and I will come back to you to report if this solved the issues or not.

**Reply by jameshulse** [reply]
Please do try it! If you have good luck then we will do the upgrade too :)

**Reply by Tofandel** [reply]
@lentex Check if it might be because of #787

**Reply by lentex** [reply]
thank you for your input @Tofandel . as far as i know all our use cases this should not be the case. but i will check
